DigitalLearningSolutions.Web/Controllers/MyAccountController.cs

```diff
@@ -248,42 +387,41 @@ DlsSubApplication dlsSubApplication
                     centreId.Value,
                     true
                 );
+                emailVerificationService.SendVerificationEmails(userAccount, unverifiedModifiedEmails);
             }
             else
             {
                 userService.UpdateUserDetails(accountDetailsData, true);
-                userService.SetCentreEmails(userId, formData.CentreSpecificEmailsByCentreId);
-            }
-
-            return this.RedirectToReturnUrl(formData.ReturnUrl, logger) ?? RedirectToAction(
-                "Index",
-                new { dlsSubApplication = dlsSubApplication.UrlSegment }
-            );
-        }
-
-        private void ValidateSingleCentreEmail(string? email, int centreId, int userId)
-        {
-            if (IsCentreSpecificEmailAlreadyInUse(email, centreId, userId))
-            {
-                ModelState.AddModelError(
-                    nameof(MyAccountEditDetailsFormData.CentreSpecificEmail),
-                    CommonValidationErrorMessages.CentreEmailAlreadyInUse
-                );
+                userService.SetCentreEmails(userAccount.Id, formData.CentreSpecificEmailsByCentreId);
+                emailVerificationService.SendVerificationEmails(userAccount, unverifiedModifiedEmails);
             }
         }
 
-        private void ValidateCentreEmailsDictionary(Dictionary<int, string?> centreEmailsDictionary, int userId)
+        private async Task<IActionResult> GetRedirectLocation(
+            UserAccount userAccount,
+            bool shouldRedirectToVerifyYourEmail,
+            bool shouldLogIntoCentrelessAccount,
+            string? returnUrl,
+            DlsSubApplication dlsSubApplication
+        )
         {
-            foreach (var (centreId, centreEmail) in centreEmailsDictionary)
+            if (shouldRedirectToVerifyYourEmail)
             {
-                if (IsCentreSpecificEmailAlreadyInUse(centreEmail, centreId, userId))
+                if (shouldLogIntoCentrelessAccount)
                 {
-                    ModelState.AddModelError(
-                        $"{nameof(MyAccountEditDetailsFormData.AllCentreSpecificEmailsDictionary)}_{centreId}",
-                        CommonValidationErrorMessages.CentreEmailAlreadyInUse
-                    );
+                    var claim = ((ClaimsIdentity)User.Identity).FindFirst("IsPersistent");
+                    var isPersistent = claim != null && Convert.ToBoolean(claim.Value);
+                    await HttpContext.Logout();
+                    await this.CentrelessLogInAsync(userAccount, isPersistent);

```

Comment:
This doesn't work because var claim is null. I wasn't very confident about it, I found the code here: https://stackoverflow.com/questions/40341591/how-to-get-is-session-ispersistent-on-asp-net-mvc

I'm also getting a `DuplicateKeyException` in `NoCachingAttribute.OnActionExecuted` sometimes when I run the code, I assume it's because of this part.